SLWL: word list (metadata) creating & editing

RickyDaMa opened the pull request:
Also adds a convenient mechanism to reduce a word list by adding a `filter` method

RickyDaMa committed:
Add WordList::filter and impl Clone for WordList

RickyDaMa committed:
Add (undocumented) builder for WordListMetadata and make it public

Make WordList::define support anything that can be transformed into WordListMetadata

RickyDaMa committed:
Add documentation & attrs to new methods

Add WordList::metadata

RickyDaMa committed:
Loosen type requirements for WordList::set_metadata

simoncozens commented on static-lang-word-lists/src/metadata.rs:
```diff
@@ -0,0 +1,170 @@
+
+impl WordListMetadata {
+    // Used by word_list!
+    // Library users should use the Builder struct
```
But why? I just don't understand what Builder adds here. You're not doing any validation on the values. It's just a way of making the API more complicated.

RickyDaMa replied:
It's to make code cleaner for creating new `WordListMetadata`s, since I'd expect that to be a more common use case than editing an existing one (though that's supported).

I don't like `WordListMetadata::new` being public because of the lack of default parameter values and named parameters in Rust. If you are just reading code without your IDE, then


```rs
let wlm = WordListMetadata::new("test", None, Some("br"));
```

doesn't tell you what the Options are, whereas the builder does:

```rs
let wlm = WordListMetadataBuilder::new("test")
    .language("br")
    .build();
```

If this was Python land I would happily just make the function signature this:

```py
def new(
    name: str,
    *,
    script: str | None = None,
    language: str | None = None,
) -> Self:
    ...
```

but unfortunately that's not possible in Rust without delving into serious black magic and/or macros.

The third option is to just have the required parameters in new, and then setters for the other fields, but then you lose the method chaining ergonomics/readability, the code is more imperative, and you have to declare `wlm` as mutable (or work around it with a code block):

```rs
let mut wlm = WordListMetadata::new("test");
wlm.set_language("br");
```

```rs
let wlm = {
    let mut wlm = WordListMetadata::new("test");
    wlm.set_language("br");
    wlm
};
```

Okay, so why not just expose the fields directly? That solves the reading-without-an-IDE problem:

```rs
let wlm = WordListMetadata {
    name: "test".into(),
    script: None,
    language: "br".into(),
};
```

but now you have to do the `into()` calls manually, and changing the fields or types of fields of the struct becomes a breaking change - maybe in future if we were to use an enum for all the scripts/languages. You can no longer add validation either.

Anyway that's a lot of words to concede that yes, it makes the API more complex, but that's intended to be for the user's benefit ultimately.

simoncozens replied:
So the point of all this extra complexity is to make it easier for people who read that one method without an IDE. Got it.

RickyDaMa committed:
Simplify: remove builder

If the builder aren't abstracting away the Cows, I might as well just give direct access to them

RickyDaMa committed:
Make WordList.metadata public
